File: spiral_matrix.cpp
#include <iostream>
using namespace std;

int main(){   
    int r = 5, c = 5;
    int ar[5][5] = {{ 1,  2,  3,  4,  5},
                    { 6,  7,  8,  9, 10},
                    {11, 12, 13, 14, 11},
                    {16, 17, 18, 19, 12},
                    {21, 22, 23, 24, 13}
    };
 //// INPUT IS GIVEN ABOVE

    int i, dir=0;
    int top = 0; // ->    # R // D 0
    int right = c-1; // I   # C // D 1
    int bottom = r-1; // <- # R // D 2
    int left = 0; // ^    # C // D 3

    while(left<=right && top<=bottom){
        if(dir==0){
            for(i=left; i<=right; i++)
                cout << ar[top][i] << " ";
            top++;
        }
        else if(dir==1){
            for(i=top; i<=bottom; i++)
                cout << ar[i][right] << " ";
            right--;
        }
        else if(dir==2){
            for(i=right; i>=left; i--)
                cout << ar[bottom][i] << " ";
            bottom--;
        }
        else if(dir==3){
            for(i=bottom; i>=top; i--)
                cout << ar[i][left] << " ";
            left++;
        }
        dir++;
        dir %= c;

        // cout << endl << "DIRECTION : " << dir <<endl;
    }

    return 0;
}

File: LEETCODE/spiral_matrix.java
//https://leetcode.com/problems/spiral-matrix/
//return all elements of the matrix in spiral order
//Input: matrix = [[1,2,3],[4,5,6],[7,8,9]]
//Output: [1,2,3,6,9,8,7,4,5]

class Solution {
    public static List<Integer> spiralOrder(int[][] matrix) {
        int startrow=0;
        int startcol=0;
        int endrow=matrix.length-1;
        int endcol=matrix[0].length-1;
          List<Integer> ls = new ArrayList<>();
       
     while(startrow<=endrow && startcol<=endcol){
         for(int j=startcol;j<=endcol;j++){
             //top
             ls.add(matrix[startrow][j]);
             
         }
            //right
             for(int i=startrow+1;i<=endrow;i++){
             
           ls.add(matrix[i][endcol]);
         }
             for(int j=endcol-1;j>=startcol;j--){
                 if(startrow==endrow){
                     break;
                 }
             //bottom
             ls.add(matrix[endrow][j]);
         }
             for(int i=endrow-1;i>=startrow+1;i--){
           //left
                 if(startcol==endcol){
                     break;
                 }
             ls.add(matrix[i][startcol]);
         }
            startrow++;
            startcol++;
            endrow--;
            endcol--;
        
    }
        return ls;
    }
    public static void main(String[] args){
        Scanner sc = new Scanner(System.in);
        
        int matrix[][]=new int[50][50];
        for(int i=0;i<matrix.length;i++){
               for(int j=0;i<matrix.length;j++){
        matrix[i][j]=sc.nextInt();
        }
        }
        System.out.println(spiralOrder(matrix));
    }
    
}
